File: harness/envelope.py
"""The one result envelope every non-Rust kyzo-bench script emits.

`kyzo_bench_harness::ResultRecord` is the Rust-side shape; this is its
Python-side equivalent for benches that time in-process library calls or
multi-service pipelines rather than subprocesses (`kuzu_baseline.py`,
`ann_baseline.py`, `driver.py` — none of these are `Rig`s, and forcing them
onto the subprocess trait would be the same modeling error kyzo#70 fixes,
relocated). Both shapes satisfy the same law
(`.claude/rules/results-data.md`): engine commit, opponent name and exact
version, dataset and its fetch hash, seed, hardware spec, date, and the
command that produced it — a result missing any of these does not land.

Every field genuinely bench-specific (Graphalytics's PR/WCC timings,
ann-benchmarks' recall/QPS curve, the kill shot's anomaly count, the
dataset identity and its fetch hash, the seed) lives in the open `metrics`
payload; it does not get its own top-level key or its own schema.
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
import time
from pathlib import Path


def hardware() -> dict[str, object]:
    cpu = "unknown"
    mem_kib = 0
    with open("/proc/cpuinfo") as f:
        for line in f:
            if line.startswith("model name"):
                cpu = line.split(":", 1)[1].strip()
                break
    with open("/proc/meminfo") as f:
        for line in f:
            if line.startswith("MemTotal:"):
                mem_kib = int(line.split()[1])
                break
    return {
        "cpu_model": cpu,
        "logical_cpus": os.cpu_count(),
        "mem_total_kib": mem_kib,
        "arch": os.uname().machine,
        "kernel": os.uname().release,
    }


def rig_commit(repo_root: Path) -> dict[str, object]:
    """The kyzo-bench commit the rig ran from — the Python analog of Rust's
    `ResultRecord::rig_commit()`. `repo_root` is any path inside the
    kyzo-bench checkout; git resolves it to the repo root itself."""
    try:
        commit = subprocess.run(
            ["git", "-C", str(repo_root), "rev-parse", "HEAD"],
            capture_output=True, text=True, check=True,
        ).stdout.strip()
        dirty = bool(subprocess.run(
            ["git", "-C", str(repo_root), "status", "--porcelain"],
            capture_output=True, text=True, check=True,
        ).stdout.strip())
        return {"commit": commit, "dirty": dirty}
    except Exception:
        return {"commit": "unknown", "dirty": True}


def today_utc() -> str:
    return time.strftime("%Y-%m-%d", time.gmtime())


def sanitize(s: str) -> str:
    return "".join(c if (c.isalnum() or c in "-.") else "_" for c in s)


def build(
    *,
    bench: str,
    story: str,
    subject: dict[str, object],
    metrics: dict[str, object],
    repo_root: Path,
    supersedes: str | None = None,
) -> dict[str, object]:
    """Assemble one envelope. `subject` is the same `{kind, name, version,
    provenance, ...}` shape Rust's `Subject` enum serializes to for every
    `kind`/`provenance.kind` Rust's `Subject`/`Provenance` enums actually
    define (`"opponent"`/`"kyzo"`; `"built_from_source"`/`"container_image"`/
    `"package"`) — build it by hand here, since there is no shared type to
    import across the language boundary. A demo comparing something Rust's
    `Provenance` has no variant for (e.g. the kill shot's four-store
    pipeline) may extend `provenance.kind` with a new, documented value
    instead of forcing a bad fit; it must not silently drop fields."""
    envelope: dict[str, object] = {
        "bench": bench,
        "story": story,
        "subject": subject,
        "rig": rig_commit(repo_root),
        "hardware": hardware(),
        "date": today_utc(),
        "command": " ".join(sys.argv),
        "metrics": metrics,
    }
    if supersedes is not None:
        envelope["supersedes"] = supersedes
    return envelope


def land(
    envelope: dict[str, object],
    results_dir: Path,
    *,
    id_: str,
    seed: int | None = None,
) -> Path:
    """Write `envelope` into `results_dir`, refusing to overwrite — the
    same append-only law `ResultRecord::land` enforces in Rust, and the
    same `{bench}--{id}--{subject}[--seed{N}]--{date}.json` naming, `id_`
    being the bench-chosen identity component (a dataset name, a workload
    id) and `seed` present only for scripts whose workload has one."""
    subject = envelope["subject"]
    subject_label = f"{subject['name']}_{subject['version']}"
    seed_part = f"--seed{seed}" if seed is not None else ""
    name = (
        f"{envelope['bench']}--{sanitize(id_)}--{sanitize(subject_label)}"
        f"{seed_part}--{envelope['date']}.json"
    )
    path = results_dir / name
    if path.exists():
        raise FileExistsError(
            f"refusing to overwrite committed result {path}: results/ is "
            "append-only; land a superseding record instead"
        )
    results_dir.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(envelope, indent=2) + "\n")
    return path


def emit(
    envelope: dict[str, object],
    *,
    land_it: bool,
    results_dir: Path,
    id_: str,
    seed: int | None = None,
) -> None:
    """The shared ending every script's `main()` calls last. Unlike Rust's
    `rig::main`, which prints *or* lands (a landed run's JSON goes to the
    result file, not stdout), this always prints — a `--land` run is
    inspectable on stdout *and* written to `results_dir` — since a bare
    Python invocation with no `--land` is the common local-dev path and
    should never require re-running to see the number. None of these three
    scripts run KyzoDB itself (that is a separate, future subject), so
    there is no engine-commit dirty gate to enforce here the way Rust's
    `rig::main` gates on `Subject::Kyzo`'s commit — `rig.dirty` is recorded
    honestly either way, landed or not."""
    print(json.dumps(envelope, indent=2))
    if land_it:
        path = land(envelope, results_dir, id_=id_, seed=seed)
        print(f"landed: {path}", file=sys.stderr)

File: harness/src/envelope.rs
//! The Rust-side twin of `harness/envelope.py`, for subject binaries that
//! sweep many measurement points inside one long-running process (an index
//! build, then N query-parameter points against it) rather than being
//! measured as N repeated subprocess invocations — `kyzo-vector-runner`,
//! `surrealdb-runner`'s vector modes, and any future binary of that shape.
//! `harness::rig`'s `Rig` trait and `ResultRecord`/`RunSet` are the right
//! fit when "run this argv, hash its output, repeat" is the whole
//! methodology; they are the wrong fit here for the same reason `envelope.py`
//! exists instead of forcing `kuzu_baseline.py` onto `Rig` — forcing this
//! shape onto `RunSet::measurements` would be recording a hash of "the
//! sweep happened" in place of the curve itself.
//!
//! Same law as the Python side (`.claude/rules/results-data.md`): engine
//! commit, opponent name and exact version, dataset identity, hardware,
//! date, and the command that produced it are structural, not optional;
//! everything bench-specific lives in the open `metrics` field.

use crate::hardware::Hardware;
use crate::subject::EngineCommit;
use serde_json::{json, Value};
use std::path::{Path, PathBuf};

#[derive(Debug, thiserror::Error)]
pub enum EnvelopeError {
    #[error(
        "refusing to overwrite committed result {path}: results/ is append-only; \
         land a superseding record instead"
    )]
    AlreadyExists { path: PathBuf },
    #[error("io error landing envelope: {0}")]
    Io(#[from] std::io::Error),
}

/// Assemble one envelope. `subject` is `kyzo_bench_harness::Subject`'s own
/// `Serialize` output — the same shape the Rust `Rig`s' `ResultRecord`
/// carries and `envelope.py`'s hand-built `subject` dict mirrors by hand
/// across the language boundary; here there is no boundary to cross, so it
/// is the real type, not a hand-copied shape of it.
pub fn build(
    bench: &str,
    story: &str,
    subject: &crate::subject::Subject,
    metrics: Value,
) -> Value {
    json!({
        "bench": bench,
        "story": story,
        "subject": subject,
        "rig": EngineCommit::capture(Path::new(".")).unwrap_or(EngineCommit {
            commit: "unknown".into(),
            dirty: true,
        }),
        "hardware": Hardware::capture(),
        "date": crate::record::ResultRecord::today_utc(),
        "command": std::env::args().collect::<Vec<_>>().join(" "),
        "metrics": metrics,
    })
}

/// Write `envelope` into `results_dir`, refusing to overwrite — identical
/// law and identical `{bench}--{id}--{subject}[--seed{N}]--{date}.json`
/// naming to `envelope.py::land`.
pub fn land(
    envelope: &Value,
    results_dir: &Path,
    id: &str,
    seed: Option<u64>,
) -> Result<PathBuf, EnvelopeError> {
    let bench = envelope["bench"].as_str().expect("bench is a string");
    let date = envelope["date"].as_str().expect("date is a string");
    let subject = &envelope["subject"];
    let subject_label = format!(
        "{}_{}",
        subject["name"].as_str().unwrap_or("unknown"),
        subject["version"].as_str().unwrap_or("unknown")
    );
    let seed_part = seed.map(|s| format!("--seed{s}")).unwrap_or_default();
    let name = format!(
        "{bench}--{}--{}{seed_part}--{date}.json",
        sanitize(id),
        sanitize(&subject_label),
    );
    let path = results_dir.join(name);
    if path.exists() {
        return Err(EnvelopeError::AlreadyExists { path });
    }
    std::fs::create_dir_all(results_dir)?;
    let text = serde_json::to_string_pretty(envelope).expect("Value always serializes");
    std::fs::write(&path, text + "\n")?;
    Ok(path)
}

/// Always print; `--land` additionally writes into `results_dir` — the same
/// "never require a re-run to see the number" rule `envelope.py::emit`
/// documents.
pub fn emit(
    envelope: Value,
    land_it: bool,
    results_dir: &Path,
    id: &str,
    seed: Option<u64>,
) -> Result<(), EnvelopeError> {
    println!("{}", serde_json::to_string_pretty(&envelope).expect("Value always serializes"));
    if land_it {
        let path = land(&envelope, results_dir, id, seed)?;
        eprintln!("landed: {}", path.display());
    }
    Ok(())
}

fn sanitize(s: &str) -> String {
    s.chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == '-' || c == '.' {
                c
            } else {
                '_'
            }
        })
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::subject::{Opponent, Provenance, Subject};

    fn envelope() -> Value {
        build(
            "vector",
            "kyzo#25",
            &Subject::Opponent(Opponent {
                name: "surrealdb".into(),
                version: "3.2.0".into(),
                provenance: Provenance::Package {
                    ecosystem: "cargo".into(),
                    package: "surrealdb".into(),
                    version: "3.2.0".into(),
                },
            }),
            json!({"curve": []}),
        )
    }

    #[test]
    fn land_refuses_overwrite() {
        let dir = std::env::temp_dir().join(format!("kb-envelope-land-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        let env = envelope();
        let first = land(&env, &dir, "sift1m", None).expect("first landing");
        assert!(first.exists());
        let err = land(&env, &dir, "sift1m", None).unwrap_err();
        assert!(matches!(err, EnvelopeError::AlreadyExists { .. }), "got: {err}");
        std::fs::remove_dir_all(&dir).unwrap();
    }

    #[test]
    fn land_filename_matches_python_convention() {
        let dir = std::env::temp_dir().join(format!("kb-envelope-name-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        let path = land(&envelope(), &dir, "sift1m", Some(42)).expect("landing");
        let name = path.file_name().unwrap().to_str().unwrap();
        assert!(name.starts_with("vector--sift1m--surrealdb_3.2.0--seed42--"), "got: {name}");
        std::fs::remove_dir_all(&dir).unwrap();
    }
}
